### ISSUE TITLE
Using the plugin without file access

### ISSUE THREAD

**Issue by gergoerdi** [ucsd-progsys/liquidhaskell#2357]
It seems that the `LiquidHaskellBoot` plugin assumes that whatever module it is processing is originally coming from a file on disk. This assumption, for example, manifests itself in `Language.Haskell.Liquid.GHC.Interface.modSummaryHsFile`, where it is assumed that the currently processed module's `ModSummary` has a non-`Nothing` value in `ml_hs_file`.

I'm working on a compiler where this assumption doesn't, in general, hold.

Is there a way to avoid that? It seems that this file is re-parsed by LH to get the refinement types. But those types are already available (in text form) in comments, which should be kept by the normal GHC parser as far as `parsedResultAction` plugins are concerned. So would it be possible for the LHB plugin to install a `parsedResultAction` hook instead of trying to get comments from the supposed original text source files?

**Comment by facundominguez** [comment]
> So would it be possible for the LHB plugin to install a parsedResultAction hook instead of trying to get comments from the supposed original text source files?

Maybe. I don't know of anyone trying that before. Relatedly, there are provisions in the code to ensure that comments are kept by the lexer (setting `Opt_KeepRawTokenStream`), but I think you saw that already.

**Comment by gergoerdi** [comment]
I've added a pull request #2367 that implements the idea in the original comment. It's not quite full file-less operation, since there's also some caching code that tries to copy files around, but I think it's a good step in the direction.

**Comment by gergoerdi** [comment]
Other than #2367, I have stumbled upon two more points where LH (or, more specifically, the `liquidhaskell-boot` plugin) tries to access source files (all findings based on aef80ef01f5445e22fc79384537a1e70420538e3):

- [ ] In `Language.Haskell.Liquid.UX.Annotate.annotate`, when `noannotations` is `False` (i.e. when annotations are enabled... as a side note, I hate flipped boolean flags...). This, of course, is easy for clients to work around by just making sure `noannotations` is always set.

  https://github.com/ucsd-progsys/liquidhaskell/blob/aef80ef01f5445e22fc79384537a1e70420538e3/liquidhaskell-boot/src/Language/Haskell/Liquid/UX/Annotate.hs#L93-L110

- [x] In `Language.Haskell.Liquid.Liquid.checkTargetInfo`. This is seemingly gated by `compileSpec`, but of course clients can't just turn off `compileSpec` without also disabling actual verification.
  
  https://github.com/ucsd-progsys/liquidhaskell/blob/aef80ef01f5445e22fc79384537a1e70420538e3/liquidhaskell-boot/src/Language/Haskell/Liquid/Liquid.hs#L36-L42

  I'm not sure what that saved file is for, but when testing locally, just removing the entire line doesn't seem to break anything, so maybe this could be controlled by a flag?

**Comment by gergoerdi** [comment]
- [x] Tangentially related, but the plugin also seems to create `.smt2` files in the `.liquid` subdirectory. These files contain the result of the LH translation from Haskell semantics to SMT problems. There is probably no good reason for these files to exist when the `Z3lib` backend is used, i.e. when SMT problems are passed to the solver in-proc with direct library calls.

**Comment by facundominguez** [comment]
> I'm not sure what that saved file is for, but when testing locally, just removing the entire line doesn't seem to break anything, so maybe this could be controlled by a flag?

Yes, I think so. These files are read when `--diffcheck` is in use. Maybe it would be fine to only write them when `--diffcheck` is enabled.

> i.e. when annotations are enabled... as a side note, I hate flipped boolean flags...

I'd be fine reversing this flag. In that case the flag would be disabled with `--annotations=False`. Perhaps it is also fine to change the default so annotations need to be requested explicitly. I never use them, but perhaps other's do. @nikivazou? @ranjitjhala?

How about removing `--noannotations` and writing them only when `--save` is enabled?

**Comment by facundominguez** [comment]
> Tangentially related, but the plugin also seems to create .smt2 files in the .liquid subdirectory.

I think it would be fine to produce these only when `--save` is active, regardless of how the smt solver is called.
